src/tests/reflection/ldtoken/modifiers.il

```diff
@@ -59,6 +59,40 @@ UIntArrayModifiedUIntArrayOK:
   ret
 MyStructPtrModifiedMyStructPtrOK:
 
+  ldtoken method int32 *()
+  stloc.0
+  ldloca 0
+  ldtoken method int32 modopt(MyModifier) *()
+
+  call instance bool valuetype [System.Runtime]System.RuntimeTypeHandle::Equals(valuetype [System.Runtime]System.RuntimeTypeHandle)
+  // All modifiers are considered as part of identity.
+  brfalse IntReturnManagedFunctionPointerModifiedOK
+  ldc.i4.6
+  ret
```

Comment:
Is this the semantic we really want? Modifiers are mostly ignored by the runtime. We made an odd choice to encode calling conventions in a spot that is normally ignored by the runtime, but it doesn't look right to upgrade all modifiers to "something the runtime cares about".

Consider this in C++/CLI:

```c++
using namespace System;

int main(array<System::String ^> ^args)
{
    typedef Int32* SimplePtr;
    typedef Int32 const* ConstPtr;
    Console::WriteLine(SimplePtr::typeid);
    Console::WriteLine(ConstPtr::typeid);
}
```

It compiles to `ldtoken    int32*` and `ldtoken    int32 modopt([mscorlib]System.Runtime.CompilerServices.IsConst)*` respectively, but the runtime behavior of both is equivalent because the modifiers don't matter to the runtime.

If we keep this semantic for function pointers where all modifiers are meaningful, it will turn function pointer types that differ by const-ness into different types at runtime. Should they be different types? Arrays aren't; pointers aren't. Why function pointers?

What are the casting rules between function pointer types that differ by (for the runtime) meaningless modifiers? How about casting rules between arrays of such function pointers?

Reply:
@MichalStrehovsky What would be the design that you would pick for this?

If we were to ignore modopts and modreqs on function pointers at runtime, `FunctionPointerParameterInfo` is going to be meaningless and always return null for the runtime reflection.

Reply:
I would expect it to work the same how it works for modifier APIs already - modifiers do show up whenever we're working with a member signature and parsing that signature. If e.g. a field is a function pointer type, I would expect modifiers to show up when reflection is parsing the field signature, same as it does for e.g. IsVolatile right now:

```
class Program
{
    public volatile int MyField;
    static void Main()
    {
        foreach (var m in typeof(Program).GetField(nameof(MyField)).GetRequiredCustomModifiers())
            Console.WriteLine(m);
    }
}
```

But I would not raise the modifiers into actual type handles (i.e. if I did a ldtoken with the IsVolatile-modified type, the IsVolatile is going to disappear).

Reply:
E.g..:

```
Console.WriteLine(typeof(Program).GetField(nameof(MyField)) == typeof(int));
```

Will print false. But that's fine, that's how reflection works everywhere else.

Reply:
> `Console.WriteLine(typeof(Program).GetField(nameof(MyField)) == typeof(int));`

This comparing `FieldInfo` with `Type`. I think you wanted to write something else.

Reply:
>  I would not raise the modifiers into actual type handles (i.e. if I did a ldtoken with the IsVolatile-modified type, the IsVolatile is going to disappear).

Is it possible to tell the difference between `static void M(delegate *<out int, void> pfn) { }` and `static void M(delegate *<ref int, void> pfn) { }` in this model?

Reply:
Yup, edited my answer.

I see what you mean. We don't have reflection facilities to actually reflection-inspect modified types - I wouldn't be able to reflection inspect a const int pointer or a const array either. We can only access modifiers in the leaf positions (i.e. modifier is directly on a field/parameter) but it's not possible to get to modifiers within the signature. I didn't realize that.

In retrospect it was probably a mistake to introduce the modifier APIs for function pointers. There hasn't been a need to inspect custom attributes in non-leaf position (as demonstrated by no API for it in 20 years) and the runtime could have just special cased things to allow callconv APIs to work in the special-purpose API.

Reply:
> Is it possible to tell the difference between static void M(delegate *<out int, void> pfn) { } and static void M(delegate *<ref int, void> pfn) { } in this model?

That comes back to the fact that reflection doesn't have deep modifier introspection APIs. I assume we wouldn't be able to tell a difference between `void M(int const * p)` and `void M(int * p)` either even though they're different methods to the runtime.

Reply:
I think we need to agree on the design goal for reflection object model.

Should the reflection object model be powerful enough to express everything that you can write in C#? In other words, should it be possible to write a C# compiler (not that anybody would want to do that) that reads and writes metadata using reflection object model? The implementation in this PR assumes that the answer is yes.

`const int pointers` or `const arrays` do not exist in C# today that makes the gap for them more acceptable. Maybe we can do a mental exercise on what we would do if we were to fix this. It may show us a better solution for the function pointers.

Reply:
> `const int pointers` or `const arrays` do not exist in C# today that makes the gap for them more acceptable.

AFAIK the C# compile will emit/respect those modifiers if it's e.g. generating a virtual override of a method that has it (i.e. wasn't written in C#). So the problem is already there, but maybe not so important.

> Maybe we can do a mental exercise on what we would do if we were to fix this. It may show us a better solution for the function pointers.

I did the mental exercise on this when I was thinking about how we could represent modifiers in the managed type system (that currently uses such bad representation that I won't even mention it).

The best I could come up with was to have a parallel way to ask about types of things raised from signatures. If we map it to the reflection API surface: e.g. for `FieldInfo` we would have a `SignatureType` in addition to the existing `FieldType` property. This `SignatureType` would either return the same results as `FieldType` (if there's no modifiers), or it would return special `System.Type` instances that represent a modified type. The special `System.Type` instance would provide APIs to query the modifiers (i.e. we would add GetRequired/GetOptionalModifiers APIs directly on `System.Type`) and get to the type being modified. One could represent any valid modifiers that way because it's just about introducing an extra `System.Type` as part of the construction from the signature.

I couldn't come up with anything better.

Reply:
Yes `foo1.GetType() == foo2.GetType()` will be `false`. The strings for both are "Program+Foo`1[System.Int32(System.Int32&)[]]"

This behavior is present in the current PR, although you may need to build in release to avoid contract checks.

Reply:
> Is this the semantic we really want? Modifiers are mostly ignored by the runtime. We made an odd choice to encode calling conventions in a spot that is normally ignored by the runtime, but it doesn't look right to upgrade all modifiers to "something the runtime cares about".

I like to think of it as:
- Today Type identify in the CLR for function pointers is based on CallKind+ReturnType+ParameterTypes.
- However, since we can't extend CallKind anymore, the newer calling conventions use modifiers (whether that was a good or bad choice, it is not relevant here).
- So we now add modifiers to Type identity for function pointers.

If the CallKind byte happended to be 2 bytes instead of 1, we probably wouldn't be having this conversation since we wouldn't have had to use modifiers for the new calling conventions and thus the way we calculate identity wouldn't have changed.

It so happens that we include *all* modifiers (opt and req for return types + parameters) in the identity in this PR but we could narrow this down to just the CallConv* mod opts on the return value. By doing this, we would also remove `FunctionPointerParameterInfo` since it wouldn't do anything more than `Type.GetFunctionPointerCallingConventions()`.

Reply:
Consider these tests:
```cs
bool b1 = typeof(delegate* unmanaged[Stdcall]<void>) == typeof(delegate* unmanaged[Fastcall]<void>);
bool b2 = typeof(delegate* unmanaged[Stdcall, SuppressGCTransition]<void>) == typeof(delegate* unmanaged[Fastcall, SuppressGCTransition]<void>);
bool b3 = typeof(delegate*<in int>) == typeof(delegate*<out int>);
```

and comparing each design option:
1) Keep the design as-is in the current PR where identity is based on all modifiers for all parameters and the return value.
2) Change to just use the CallConv* (mod opts on the return value).
3) Revert back to where the runtime is today (no modifiers).
4) Return whatever Type instance we want when doing reflection; does not map 1:1 with runtime type descriptors\handles which remain unchanged.

| Test | Option 1 | Option 2 | Option 3 | Option 4
| --------------- | --------------- | --------------- | --------------- | --------------- |
| b1 | false | false | undefined | TBD
| b2 | false | false | undefined | TBD
| b3 | false | true | undefined | TBD

undefined = this can be encoded via CallKind or modifiers. C# has specific rules, but not everyone may do the same.

Reply:
cc @jaredpar: what do you think about equality for function pointers?

Reply:
I think the ability to get the modifiers and calling conventions in general will be most useful for MetadataLoadContext scenarios, which of course use these APIs as well.

Reply:
> I think the ability to get the modifiers and calling conventions in general will be most useful for MetadataLoadContext scenarios, which of course use these APIs as well.

Yes, that matches what Jan wrote above - the ability to deeply inspect the modifiers is good because it lets one write more powerful tools. But we currently don't have any other reflection APIs to deeply inspect modified types and Jan's question is how we would go about address _that_ and whether it would match what we're doing here for function pointer types. The problem is that there are other places where a compiler can legally place a custom modifier that we currently don't surface - the example I wrote in C++/CLI that started this is one (the `IsConst` modifier C++/CLI places on pointers to const things). How would we go about fixing that? Would we surface `const int` as a separate type with a type handle and all? Or would we choose a different avenue?

`const int` type is currently indistinguishable from `int` to both runtime reflection and `MetadataLoadContext` - we are actually doing a funny thing here because a function pointer to a method that takes a `int const *` will be represented by a different type than one that takes a `int *` but if you structurally try to compare why the function pointers are different with reflection APIs, they'll both be the same because we do drop the modifier after the pointer.

Giving the modifiers a meaning also means we have a breaking change in:

```csharp
object arr1 = new delegate*<out int, int>[1];
var arr2 = (delegate*<ref int, int>[])arr1;
```

These two arrays are currently castable. They won't be after this change. The C# language team would rejoice at this breaking change because that's the C# semantic. But it's not the runtime semantic elsewhere. E.g.:

```csharp
class Program
{
    delegate void Del(ref int x);
    public static void Method(in int x) { }
    unsafe static void Main()
    {
        // We allow creating delegates with a mismatch in ref/in/out because these are all just byrefs to the runtime
        var d = typeof(Program).GetMethod(nameof(Method)).CreateDelegate<Del>();
        int x = 123;
        d(ref x);
    }
}
```